src/Microsoft.ML.TimeSeries/SrCnnBatchAnomalyDetection.cs

```diff
@@ -0,0 +1,778 @@
+            if (batchSize == -1)
+            {
+                _batchSize = (int)input.GetRowCount();
+            }
+            else
+            {
+                _batchSize = batchSize;
+            }
+
+            env.CheckUserArg(threshold >= 0 && threshold <= 1, nameof(threshold), "Must be in [0,1].");
+            env.CheckUserArg(detectMode == SrCnnDetectMode.AnomalyOnly
+                || detectMode == SrCnnDetectMode.AnomalyAndExpectedValue
+                || detectMode == SrCnnDetectMode.AnomalyAndMargin, nameof(detectMode), "Invalid detectMode");
+
+            if (detectMode.Equals(SrCnnDetectMode.AnomalyOnly))
+            {
+                _outputLength = AnomalyOnlyOutputLength;
+                _modler = new SrCnnEntireModeler(threshold, sensitivity, detectMode, _outputLength);
+            }
+            else if (detectMode.Equals(SrCnnDetectMode.AnomalyAndMargin))
+            {
+                env.CheckUserArg(sensitivity >= 0 && sensitivity <= 100, nameof(sensitivity), "Must be in [0,100].");
+                _outputLength = AnomalyAndMarginOutputLength;
+                _modler = new SrCnnEntireModeler(threshold, sensitivity, detectMode, _outputLength);
+            }
+            else if (detectMode.Equals(SrCnnDetectMode.AnomalyAndExpectedValue))
+            {
+                _outputLength = AnomalyAndExpectedValueOutputLength;
+                _modler = new SrCnnEntireModeler(threshold, sensitivity, detectMode, _outputLength);
+            }
+
+            SchemaBindings = new Bindings(input.Schema, inputColumnName, outputColumnName, new VectorDataViewType(NumberDataViewType.Double, _outputLength));
+        }
+
+        protected override ColumnBindingsBase SchemaBindings { get; }
+
+        protected override Delegate[] CreateGetters(DataViewRowCursor input, Batch currentBatch, bool[] active)
+        {
+            if (!SchemaBindings.AnyNewColumnsActive(x => active[x]))
+                return new Delegate[1];
+            return new[] { currentBatch.CreateGetter(input, _inputColumnName) };
+        }
+
+        protected override Batch InitializeBatch(DataViewRowCursor input) => new Batch(_batchSize, _outputLength, _modler);
+
+        protected override Func<bool> GetIsNewBatchDelegate(DataViewRowCursor input)
+        {
+            return () => input.Position % _batchSize == 0;
+        }
+
+        protected override Func<bool> GetLastInBatchDelegate(DataViewRowCursor input)
+        {
+            return () => (input.Position + 1) % _batchSize == 0;
+        }
+
+        protected override ValueGetter<double> GetLookAheadGetter(DataViewRowCursor input)
+        {
+            return input.GetGetter<double>(input.Schema[_inputColumnName]);
+        }
+
+        protected override Func<int, bool> GetSchemaBindingDependencies(Func<int, bool> predicate)
+        {
+            return (SchemaBindings as Bindings).GetDependencies(predicate);
```

Comment:
>SchemaBindings [](start = 20, length = 14)

Suggestion: Add a private field of type `Bindings`:
```
private readonly Bindings _bindings;
```
and then the override of `SchemaBindings` would be:
```
public override ColumnBindingsBase SchemaBindings => _bindings;
```
and this method would return `_bindings.GetDependencies(predicate)`.

Reply:
Fixed.

---
In reply to: [426410915](https://github.com/dotnet/machinelearning/pull/5135#discussion_r426410915) [](ancestors = 426410915)